library/core/src/primitive_docs.rs

```diff
@@ -268,6 +268,52 @@ mod prim_bool {}
 /// [`Debug`]: fmt::Debug
 /// [`default()`]: Default::default
 ///
+/// # Never type fallback
+///
+/// When the compiler sees a value of type `!` it implicitly inserts a coercion (if possible)
+/// to allow type checker to infer any type:
+///
+/// ```rust,ignore (illustrative-and-has-placeholders)
+/// // this
+/// let x: u8 = panic!();
+///
+/// // is (essentially) turned by the compiler into
+/// let x: u8 = absurd(panic!());
+///
+/// // where absurd is a function with the following signature
+/// // (it's sound, because `!` always marks unreachable code):
+/// fn absurd<T>(_: !) -> T { ... }
+// FIXME: use `core::convert::absurd` here instead, once it's merged
+/// ```
+///
+/// While it's convenient to be able to use non-diverging code in one of the branches (like
+/// `if a { b } else { return }`) this could lead to compilation errors:
```

Comment:
As is, the sentence could be read like “using non-diverging code in a branch could lead to compilation errors”, so let's make it more explicit:

```suggestion
/// While this coercion gives us the convenience of being able to use non-diverging code in
/// one of the branches (like `if a { b } else { return }`), it could lead to compilation
/// errors:
```

Reply:
Maybe just replace this with:
```
This can lead to compilation errors if the type cannot be inferred.
```

Reply:
I agree with @jongiddy, I don't think this section needs to justify why the coercion is convenient, it only needs to explain the fallback

Reply:
The reason I added this mention is to prevent a question like "but why does compiler does that, if it causes problems?". Why do you think it's unnecessary?

Reply:
The text before the fallback section gives motivating examples of `!` being coerced to provide a consistent type. The explanation here doesn't explain why it is convenient, so the reader needs to be familiar with those previous sections to understand this sentence. And there is an ambiguity in the phrasing that suggests that the `if` example can lead to compilation errors.

Reply:
Yeah, we can probably land this doc without this additional motivation and add more when someone asks.

Reply:
Sure.

```suggestion
/// This can lead to compilation errors if the type cannot be inferred:
```